pkg/services/ngalert/eval/eval.go

```diff
@@ -0,0 +1,196 @@
+	// that evaluated to false.
+	Normal state = iota
+
+	// Alerting is the eval state for an alert instance condition
+	// that evaluated to false.
+	Alerting
+)
+
+func (s state) String() string {
+	return [...]string{"Normal", "Alerting"}[s]
+}
+
+// IsValid checks the condition's validity.
+func (c Condition) IsValid() bool {
+	// TODO search for refIDs in QueriesAndExpressions
+	return len(c.QueriesAndExpressions) != 0
+}
+
+// AlertExecCtx is the context provided for executing an alert condition.
+type AlertExecCtx struct {
+	AlertDefitionID int64
+	SignedInUser    *models.SignedInUser
+
+	Ctx context.Context
+}
+
+// Execute runs the Condition's expressions or queries.
+func (c *Condition) Execute(ctx AlertExecCtx, fromStr, toStr string) (*ExecutionResults, error) {
+	result := ExecutionResults{}
+	if !c.IsValid() {
+		return nil, fmt.Errorf("invalid conditions")
+	}
+
+	pbQuery := &pluginv2.QueryDataRequest{
+		PluginContext: &pluginv2.PluginContext{
+			// TODO: Things probably
+		},
+		Queries: []*pluginv2.DataQuery{},
+	}
+
+	for i := range c.QueriesAndExpressions {
+		q := c.QueriesAndExpressions[i]
+		model, err := q.getModel()
+		if err != nil {
+			return nil, fmt.Errorf("failed to get query model: %w", err)
+		}
+		intervalMS, err := q.getIntervalMS()
+		if err != nil {
+			return nil, fmt.Errorf("failed to retrieve intervalMs from the model: %w", err)
+		}
+
+		maxDatapoints, err := q.getMaxDatapoints()
+		if err != nil {
+			return nil, fmt.Errorf("failed to retrieve maxDatapoints from the model: %w", err)
+		}
+
+		pbQuery.Queries = append(pbQuery.Queries, &pluginv2.DataQuery{
+			Json:          model,
+			IntervalMS:    intervalMS,
+			RefId:         q.RefID,
+			MaxDataPoints: maxDatapoints,
+			QueryType:     q.QueryType,
+			TimeRange:     q.RelativeTimeRange.toTimeRange(time.Now()),
+		})
+	}
+
+	tw := plugins.Transform
+	pbRes, err := tw.TransformClient.TransformData(ctx.Ctx, pbQuery, tw.Callback)
+	if err != nil {
+		return &result, err
+	}
+
+	for refID, res := range pbRes.Responses {
+		if refID != c.RefID {
+			continue
+		}
+		df := tsdb.NewEncodedDataFrames(res.Frames)
+		result.Results, err = df.Decoded()
+		if err != nil {
+			result.Error = err
+			return &result, err
+		}
+	}
+
+	if len(result.Results) == 0 {
+		err = fmt.Errorf("no GEL results")
+		result.Error = err
+		return &result, err
+	}
+
+	return &result, nil
+}
+
+// EvaluateExecutionResult takes the ExecutionResult, and returns a frame where
+// each column is a string type that holds a string representing its state.
+func EvaluateExecutionResult(results *ExecutionResults) (Results, error) {
+	evalResults := make([]result, 0)
+	labels := make(map[string]bool)
+	for _, f := range results.Results {
+		rowLen, err := f.RowLen()
+		if err != nil {
+			return nil, fmt.Errorf("unable to get frame row length: %w", err)
+		}
+		if rowLen > 1 {
+			return nil, fmt.Errorf("invalid frame %q: row length: %v", f.Name, rowLen)
```

Comment:
I hit this one pointing at a time series queries instead of the reduction query:

```
{
  "error": "invalid frame \"\": row length: 9",
  "message": "Failed to evaluate results"
}
```

Took me a bit to figure out where/why.  I think maybe we need to wrap this with an generic error explaining that the results returned were not in a valid format. Also maybe referencing the condition/refId might help. 

With this specific one, we should probably also state the expected length in the details of the error (0 or 1)